### PR TITLE
Remove or refactor evaluation extension function errors

### LINKED ISSUE #193: Improvements to ExtensionsError variants representing internal invariant violations
### Category

User level API changes

### Describe the feature you'd like to request

Several variants of `ExtensionsError` -- `HasNoType`, `FuncMultiplyDefined`, and `MultipleConstructorsSameSignature` -- represent internal invariant violations more than they represent something the user did wrong.  In the current world where we control all the language extensions, these should never be thrown, and if they are thrown, it represents a bug in our code.

We should mark these as internal errors -- perhaps by making a struct `InternalError` (or some other name) and having that be a variant of `ExtensionsError`.  If possible, we should also reorganize `ExtensionsError` and/or `EvaluationError` so that these errors are (a) caught as early as possible, eg before `is_authorized()`, and (b) not part of `EvaluationError`, so that consumers of `EvaluationError` don't have to worry about matching on these.

We want to catch these as early as possible, but we also don't want to do unnecessary expensive checks on the hot path.  One option is to do it when constructing an `Extensions` object, if that is not on the hot path or can be moved off the hot path.

### Describe the solution you'd like

As above.

### Describe alternatives you've considered

Removing these variants entirely in favor of some other mechanism or refactor that makes those errors impossible.  This might be very difficult without restructuring our extensions mechanism.

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
## Description of changes

Update extension function errors in evaluation to fix #193. A few more changes for validation will come in a second PR.

Changes:

* Move `FuncMultiplyDefinedError` into a new error enum `ExtensionInitializationError` which is only returned when constructing extension and never in evaluation.
* Move `MultipleConstructorsSameSignature` into the `ExtensionInitializationError` enum 
* replace `HasNoType` with the existing variant `UnknownInsufficientTypeInfoError` a `None` return type on an extension function could only happen when that extension function was `unknown`, so these errors were effectively the same. Adjust some names an comments to make it more clear what `None` in the returns means.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [ ] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

### CODE REVIEW

File: cedar-policy-core/src/ast/extension.rs [hunk 1]
```diff
@@ -301,7 +303,7 @@ impl ExtensionFunction {
     ///
     /// If any given argument type is not constant (function works with multiple
     /// `SchemaType`s) then this will be `None` for that argument.
```
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/extension.rs:305]: Comment needs to be removed here as well

File: cedar-policy-core/src/entities/json/schema_types.rs [hunk 2]
```diff
@@ -319,6 +312,40 @@ pub struct HeterogeneousSetError {
...
+pub struct UnknownInsufficientTypeInfoError {
+    /// `Unknown` which has insufficient type information
+    unknown: Option<Unknown>,
+}
+
+impl std::fmt::Display for UnknownInsufficientTypeInfoError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "cannot compute type because of insufficient type information for"
```
Comment (cdisselkoen) [inline cedar-policy-core/src/entities/json/schema_types.rs:328]: I think you need a trailing space
```suggestion
            "cannot compute type because of insufficient type information for "
```

File: cedar-policy-core/src/entities/json/schema_types.rs [hunk 3]
```diff
@@ -319,6 +312,40 @@ pub struct HeterogeneousSetError {
...
+/// residual which is not just a single `Unknown`). For now, we do not
+/// attempt to compute the [`SchemaType`] in these cases, and just return
+/// this error.
+#[derive(Debug, Diagnostic, Error)]
+#[error("cannot compute type of nontrivial residual `{residual}`")]
+pub struct NontrivialResidualError {
+    /// Nontrivial residual which we were trying to compute the
+    /// [`SchemaType`] of
+    residual: Box<Expr>,
+}
```
Comment (cdisselkoen) [inline cedar-policy-core/src/entities/json/schema_types.rs:347]: bonus points for a `Diagnostic` implementation to take `labels()` and `source_code()` from the expr's `Loc`.  There are other errors you can copy-paste this implementation from (or make a macro)

File: cedar-policy-core/src/entities.rs [hunk 4]
```diff
@@ -174,7 +174,7 @@ impl Entities {
             // We do this before adding the actions, because we trust the
             // actions were already validated as part of constructing the
             // `Schema`
-            let checker = EntitySchemaConformanceChecker::new(schema, extensions);
+            let checker = EntitySchemaConformanceChecker::new(schema, extensions.clone());
```
Comment (cdisselkoen) [inline cedar-policy-core/src/entities.rs:177]: is it possible for `EntitySchemaConformanceChecker` to hold an `&Extensions` instead?  (I do recognize that `Extensions` is still pretty cheap to clone, so this is a minor nit)

File: cedar-policy-core/src/entities/json/context.rs [hunk 5]
```diff
@@ -82,12 +82,12 @@ impl<'e, 's, S: ContextSchema> ContextJsonParser<'e, 's, S> {
         &self,
         json: serde_json::Value,
     ) -> Result<Context, ContextJsonDeserializationError> {
-        let vparser = ValueParser::new(self.extensions);
+        let vparser = ValueParser::new(self.extensions.clone());
```
Comment (cdisselkoen) [inline cedar-policy-core/src/entities/json/context.rs:85]: is it possible for `ValueParser` to hold an `&Extensions` instead?  (I do recognize that `Extensions` is still pretty cheap to clone, so this is a minor nit)

File: cedar-policy-core/src/entities/json/context.rs [hunk 6]
```diff
@@ -82,12 +82,12 @@ impl<'e, 's, S: ContextSchema> ContextJsonParser<'e, 's, S> {
...
         let expected_ty = self.schema.map(|s| s.context_type());
         let rexpr = vparser.val_into_restricted_expr(json, expected_ty.as_ref(), || {
             JsonDeserializationErrorContext::Context
         })?;
-        Context::from_expr(rexpr.as_borrowed(), self.extensions)
+        Context::from_expr(rexpr.as_borrowed(), self.extensions.clone())
```
Comment (cdisselkoen) [inline cedar-policy-core/src/entities/json/context.rs:90]: Is it possible for `Context::from_expr()` to take an `&Extensions` instead?

File: cedar-policy-core/src/evaluator.rs [hunk 7]
```diff
@@ -172,7 +172,7 @@ impl<'e> RestrictedEvaluator<'e> {
                 match split(args) {
                     Either::Left(values) => {
                         let values : Vec<_> = values.collect();
-                        let efunc = self.extensions.func(fn_name).map_err(|err| err.with_maybe_source_loc(expr.source_loc().cloned()))?;
+                        let efunc = self.extensions.func(fn_name)?;
```
Comment (cdisselkoen) [inline cedar-policy-core/src/evaluator.rs:175]: why this change?

File: cedar-policy-core/src/evaluator.rs [hunk 8]
```diff
@@ -541,10 +541,7 @@ impl<'e> Evaluator<'e> {
                 match split(args) {
                     Either::Left(vals) => {
                         let vals: Vec<_> = vals.collect();
-                        let efunc = self
-                            .extensions
-                            .func(fn_name)
-                            .map_err(|err| err.with_maybe_source_loc(loc.cloned()))?;
+                        let efunc = self.extensions.func(fn_name)?;
```
Comment (cdisselkoen) [inline cedar-policy-core/src/evaluator.rs:544]: why this change?

File: cedar-policy-core/src/extensions.rs [hunk 9]
```diff
@@ -23,49 +23,152 @@ pub mod ipaddr;
...
 ///
 /// Clone is cheap for this type.
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone)]
 pub struct Extensions<'a> {
     /// the actual extensions
     extensions: &'a [Extension],
+    /// All extension functions, collected from every extension used to
+    /// construct this object.  Built ahead of time so that we know during
+    /// extension function lookup that at most one extension functions exists
```
Comment (cdisselkoen) [inline cedar-policy-core/src/extensions.rs:77]:
```suggestion
    /// extension function lookup that at most one extension function exists
```

File: cedar-policy-core/src/extensions.rs [hunk 10]
```diff
@@ -124,105 +214,75 @@ impl<'a> Extensions<'a> {
     /// `Err` is returned in the case that multiple constructors have that signature.
```
Comment (cdisselkoen) [inline cedar-policy-core/src/extensions.rs:214]: Comment needs updating here too. (and the polymorphic comment no longer applies)

File: cedar-policy-core/src/extensions.rs [hunk 11]
```diff
@@ -233,7 +293,7 @@ pub mod extension_function_lookup_errors {
     // Don't make fields `pub`, don't make breaking changes, and use caution
     // when adding public methods.
     #[derive(Debug, PartialEq, Eq, Clone, Error)]
-    #[error("extension function `{name}` does not exist")]
+    #[error("internal invariant violated. Extension function `{name}` does not exist")]
```
Comment (cdisselkoen) [inline cedar-policy-core/src/extensions.rs:296]: Might be worth commenting why this is an internal invariant violation, and why you won't get this error from a policy that calls `foo()` (in either JSON or Cedar format), or entity JSON data that calls `foo()`

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-core/src/evaluator.rs (hunk 7 above)
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/evaluator.rs:175]: `Name`s carry a source location, so I update the error to use that

File: cedar-policy-core/src/entities.rs (hunk 4 above)
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/entities.rs:177]: Yes. I'm confident I could get rid of all the clones and just do refs everywhere, but it creates a bunch of changes, so I didn't go through with it

Comment (cdisselkoen) [reply cedar-policy-core/src/entities.rs:177]: Sounds fine

File: cedar-policy-core/src/extensions.rs (hunk 11 above)
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/extensions.rs:294]: Looked at this closer, and the supposed invariant is violated when parsing entity data:

```json
    {
        "uid": {  "type":"User", "id":"bob" },
        "attrs": {
            "score": { "__extn": { "fn": "foo", "arg": "0.899" }}
        },
        "parents": []
    }
```
```
  × failed to parse entities from file test.json
  ├─▶ error during entity deserialization
  ╰─▶ failed to evaluate attribute `score` of `User::"bob"`: internal invariant violated. Extension function `foo` does not exist
```

Not a Cedar bug because we're getting an error exactly where we want one. Just means this error variant can occur legitimately without some more refactoring

Comment (john-h-kastner-aws) [reply cedar-policy-core/src/extensions.rs:294]: IMO not worth restructuring just so we can say this is an internal invariant violation.

File: cedar-policy-core/src/entities/json/schema_types.rs (hunk 3 above)
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/entities/json/schema_types.rs]: done

File: cedar-policy-core/src/extensions.rs (hunk 11 above)
Comment (cdisselkoen) [reply cedar-policy-core/src/extensions.rs]: Makes sense to me

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy-core/src/entities.rs (hunk 4 above)
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/entities.rs:177]: not so bad actually #1100
